src/backgrounddownloader.cpp

```diff
@@ -0,0 +1,152 @@
+        emit(confirmStartDownload(bookId, QString::fromStdString(download->getDid())));
+    } catch (std::exception& e) {
+        // Download failed, ignore
+        std::cerr << "got exception " << e.what() << '\n';
+    }
+}
+
+// Cancel the download if it is running, then remove from the m_status map
+void BackgroundDownloader::completeDownload(const QString& did)
+{
+    kiwix::Download* download = mp_downloader->getDownload(did.toStdString());
+
+    m_rwlock.lockForWrite();
+    QMap<std::string, std::string> map = m_status.value(did.toStdString());
+    m_status.remove(did.toStdString());
+    download->cancelDownload();
+    m_rwlock.unlock();
+}
+
+void BackgroundDownloader::pauseDownload(const QString& did)
+{
+    kiwix::Download* download = mp_downloader->getDownload(did.toStdString());
+    if (download->getStatus() == kiwix::Download::K_ACTIVE) {
+        m_rwlock.lockForWrite();
+        download->pauseDownload();
+        m_rwlock.unlock();
+    }
+}
+
+void BackgroundDownloader::resumeDownload(const QString& did)
+{
+    kiwix::Download* download = mp_downloader->getDownload(did.toStdString());
+    if (download->getStatus() == kiwix::Download::K_PAUSED) {
+        m_rwlock.lockForWrite();
+        download->resumeDownload();
+        m_rwlock.unlock();
+    }
+}
+
+void BackgroundDownloader::cancelDownload(const QString& bookId, const QString& did)
+{
+    qInfo() << "canceling download for book = " << bookId;
+    kiwix::Download* download = mp_downloader->getDownload(did.toStdString());
+
+    m_rwlock.lockForWrite();
+    QMap<std::string, std::string> map = m_status.value(did.toStdString());
+    m_status.remove(did.toStdString());
+    download->cancelDownload();
+    m_rwlock.unlock();
+
+    emit(confirmCancelDownload(bookId, QString::fromStdString(map.value("path"))));
+}
+
+void BackgroundDownloader::updateStatus()
+{
+    for (auto& did : mp_downloader->getDownloadIds()) {
+        qInfo() << "updating status for did " << did.c_str();
+        kiwix::Download* download = mp_downloader->getDownload(did);
+        download->updateStatus(true);
+
+        m_rwlock.lockForWrite();
+        auto status = m_status.value(did);
+        status.insert("id", download->getDid());
+        status.insert("followed_by", download->getFollowedBy());
+        status.insert("path", download->getPath());
+        status.insert("totalLength", std::to_string(download->getTotalLength()));
+        status.insert("completedLength", std::to_string(download->getCompletedLength()));
+        status.insert("downloadSpeed", std::to_string(download->getDownloadSpeed()));
+        status.insert("verifiedLength", std::to_string(download->getVerifiedLength()));
+        status.insert("status", convertStatusResult(download->getStatus()));
+        m_status.insert(did, status);
+        m_rwlock.unlock();
+    }
+    int mapSize = 0;
+    for (auto it = m_status.begin(); it != m_status.end(); ++it) {
+        mapSize++;
+    }
```

Comment:
This iterator loop was just debugging a problem with an unexpected map size being reported. Its equivalent to `m_status.size()`. 

I found the download map was bigger than expected because `mp_downloader->getDownloadIds()` never forgets downloads. Even though the map entry is removed during cancel, it gets added back again because the old download ID is always returned.